sungrow/integrationpluginsungrow.cpp

```diff
@@ -0,0 +1,375 @@
+            thing->setStateValue("connected", success);
+
+            foreach (Thing *childThing, myThings().filterByParentId(thing->id())) {
+                childThing->setStateValue("connected", success);
+            }
+
+            if (!success) {
+                // Try once to reconnect the device
+                sungrowConnection->reconnectDevice();
+            } else {
+                qCInfo(dcSungrow()) << "Connection initialized successfully for" << thing;
+                sungrowConnection->update();
+            }
+        });
+
+        connect(sungrowConnection, &SungrowModbusTcpConnection::updateFinished, thing, [=](){
+            qCDebug(dcSungrow()) << "Updated" << sungrowConnection;
+
+            if (myThings().filterByParentId(thing->id()).filterByThingClassId(sungrowMeterThingClassId).isEmpty()) {
+                qCDebug(dcSungrow()) << "There is no meter set up for this inverter. Creating a meter for" << thing << sungrowConnection->modbusTcpMaster();
+                ThingClass meterThingClass = thingClass(sungrowMeterThingClassId);
+                ThingDescriptor descriptor(sungrowMeterThingClassId, meterThingClass.displayName() + " " + sungrowConnection->serialNumber(), QString(), thing->id());
+                emit autoThingsAppeared(ThingDescriptors() << descriptor);
+            }
+
+            // Check if a battery is connected to this Sungrow inverter
+            if (sungrowConnection->batteryType() != SungrowModbusTcpConnection::BatteryTypeNoBattery &&
+                    myThings().filterByParentId(thing->id()).filterByThingClassId(sungrowBatteryThingClassId).isEmpty()) {
+                qCDebug(dcSungrow()) << "There is a battery connected but not set up yet. Creating a battery.";
+                ThingClass batteryThingClass = thingClass(sungrowBatteryThingClassId);
+                ThingDescriptor descriptor(sungrowBatteryThingClassId, batteryThingClass.displayName() + " " + sungrowConnection->serialNumber(), QString(), thing->id());
+                emit autoThingsAppeared(ThingDescriptors() << descriptor);
+            }
+
+            // Update inverter states
+            thing->setStateValue(sungrowInverterTcpCurrentPowerStateTypeId, sungrowConnection->totalPVPower());
+            thing->setStateValue(sungrowInverterTcpTemperatureStateTypeId, sungrowConnection->inverterTemperature());
+            thing->setStateValue(sungrowInverterTcpFrequencyStateTypeId, sungrowConnection->gridFrequency());
+            thing->setStateValue(sungrowInverterTcpTotalEnergyProducedStateTypeId, sungrowConnection->totalPVGeneration());
+
+            // Update the meter if available
+            Thing *meterThing = getMeterThing(thing);
+            if (meterThing) {
+                auto runningState = sungrowConnection->runningState();
+                qCInfo(dcSungrow()) << "Power generated from PV:" << (runningState & (0x1 << 0) ? "true" : "false");
```

Comment:
Such details of the integration could be handled using `qCDebug` since they are only interesting while debugging/development.